File: algorithm/杭电ACM/杭电ACM_C++/杭电题/2116.cpp
////////////////////System Comment////////////////////
////Welcome to Hangzhou Dianzi University Online Judge
////http://acm.hdu.edu.cn
//////////////////////////////////////////////////////
////Username: devil_momo
////Nickname: Lucifer
////Run ID: 
////Submit time: 2008-09-06 13:36:38
////Compiler: Visual C++
//////////////////////////////////////////////////////
////Problem ID: 2116
////Problem Title: 
////Run result: Accept
////Run time:0MS
////Run memory:0KB
//////////////////System Comment End//////////////////
#include <iostream>
using namespace std;
int main()
{
    __int64 a,b,max,min;
    int n,f;
    while(scanf("%d",&n)==1)
    {
        max=(__int64)1<<(n-1);
		min=-max;
		max-=1;
        scanf("%I64d%I64d",&a,&b);
		f=1;
		if(n==64)
		{
			if(a<0&&b<0)
			{
				if(a+b>=0)
				f=0;
			}
			else if(a>=0&&b>=0)
			{
				if(a+b<=0)
					f=0;
			}
		}
		else
		{
			if(a+b>max||a+b<min)
				f=0;
		}
        puts(f?"WaHaHa":"Yes");
    }
    return 0;
}

File: algorithm/杭电ACM/ambition0109/2116.java
////////////////////System Comment////////////////////
////Welcome to Hangzhou Dianzi University Online Judge
////http://acm.hdu.edu.cn
//////////////////////////////////////////////////////
////Username: ambition0109
////Nickname: Ambition
////Run ID: 
////Submit time: 2010-11-20 19:49:34
////Compiler: Java
//////////////////////////////////////////////////////
////Problem ID: 2116
////Problem Title: 
////Run result: Accept
////Run time:93MS
////Run memory:6300KB
//////////////////System Comment End//////////////////
import java.util.Scanner;
import java.math.BigInteger;

public class Main {
	
	public static void main(String arga[]){
		Scanner in=new Scanner(System.in);
		
		while(in.hasNext()){
			int n=in.nextInt();
			BigInteger tmp=BigInteger.ONE;
			BigInteger a=in.nextBigInteger();
			BigInteger b=in.nextBigInteger();
			BigInteger sum=a.add(b);
			tmp=tmp.shiftLeft(n-1);
			if(sum.compareTo(BigInteger.ZERO)>=0){
				sum=sum.add(BigInteger.ONE);
			}else{
				sum=sum.abs();
			}
		//	System.out.println(sum+"  "+tmp);
			if(sum.compareTo(tmp)>0){
				System.out.println("Yes");
			}else{
				System.out.println("WaHaHa");
			}
		}
	}
}
